src/http/h1/parse.rs

```diff
@@ -35,15 +36,24 @@ impl Http1Transaction for ServerTransaction {
                 trace!("Request.parse Complete({})", len);
                 let mut headers = Headers::with_capacity(req.headers.len());
                 let slice = buf.slice(len);
+                let path = req.path.unwrap();
+                let path_start = path.as_ptr() as usize -
+                    unsafe { MemStr::from_utf8_unchecked(slice.clone()) }.as_str()
+                                                                         .as_ptr() as usize;
+                let path_end = path_start + path.len();
+                let path = slice.slice(path_start..path_end);
                 headers.extend(HeadersAsMemSliceIter {
                     headers: req.headers.iter(),
                     slice: slice,
                 });
+                // path was found to be utf8 by httparse
+                let path = unsafe { MemStr::from_utf8_unchecked(path) };
+
                 Some((MessageHead {
                     version: if req.version.unwrap() == 1 { Http11 } else { Http10 },
                     subject: RequestLine(
                         try!(req.method.unwrap().parse()),
-                        try!(req.path.unwrap().parse())
+                        try!(::uri::Uri::from_str(path.as_str())),
```

Comment:
This goes through `Uri::from_str` -> `InternalUri::from(&str)` -> `str.to_owned()`, which means we're still copying the bytes here. We could instead introduce a new function in the `uri` module like so:

```rust
// in uri.rs
pub fn from_mem_str(s: MemStr) -> Result<Uri, Error> {
    Uri::new(InternalUri::from(s))
}
```

And then we can use `::uri::from_mem_str(path)`. Since the `uri` module isn't public, and only the `Uri` type is re-exported outside of the crate, we can keep the implementation details private.